docs/keps/20200504-cleanup-fed-resources.md

```diff
@@ -0,0 +1,113 @@
+---
+kep-number: 0
+short-desc: Cleanup federated resource on unjoin
+title: Cleanup Federated Resource
+authors:
+  - "@jr0d"
+reviewers:
+  - "@jimmidyson"
+  - "@hectorj2f"
+approvers:
+  - "@jimmidyson"
+editor: TBD
+creation-date: 2020-05-18
+last-updated: 2020-05-04
+status: provisional
+---
+
+# Cleanup federated resources
+
+## Table of Contents
+
+* [Cleanup Federated Resources](#cleanup-federated-resources)
+  * [Table of Contents](#table-of-contents)
+  * [Summary](#summary)
+  * [Motivation](#motivation)
+    * [User Story](#user-stroy)
+    * [Goals](#goals)
+  * [Proposals](#proposal)
+    * [Implementation Details](#implementation-details)
+      * [Best Effort Strategy](#best-effort-strategy)
+      * [Required Strategy](#required-strategy)
+    * [Addition ClusterConditionTypes (optional)](#additional-clusterconfigtypes-optional)
+      * [Unjoining ClusterConfigurationType](#unjoining-clusterconfigurationtype)
+      * [Failed ClusterConfigurationType](#failed-clusterconfigurationtype)
+  * [Risks](#risks)
+    * [Adding Cluster Conditions](#adding-cluster-conditions)
+
+This document proposes a mechanism to specify that a federated resource should be removed from a managed cluster when leaving federation.
+
+## Motivation
+
+In practice, federation is often used to join clusters to some higher level centralized management plane, identity management for example. To achieve this, federation is used to propagate a configuration state to clusters which overrides an existing configuration state. When a cluster is unjoined, the federated configuration becomes invalid requiring the operator to manually remove the conflicting resources.
+
+### User Story
+
+As a cluster operator using kubefed, I would like to create resources which will automatically be removed from managed clusters during unjoin.
+
+### Goals
+
+* Define how users can indicate that a federated resource should be removed when a cluster leaves federation
+* Define removal strategies
+
+## Proposals
+
+The APIResource struct should be updated to contain a `RemoveStrategy` member. `RemoveStrategy` will initially support two values, `BestEffort` and `Required`.
+
+  * The `BestEffort` strategy will **not** prevent a cluster from unjoining when there are errors removing the resource.
+  * The `Required` strategy will halt the unjoin process and set the KubefedCluster resource in a failed state when encountering errors. Unjoining is blocked until all resources with the `Required` `RemoveStrategy` have been removed successfully.
+
+By default, `RemoveStrategy` will be nil on all federated types. A user must explicitly set a removal strategy on the resource they are creating. Types without a removal strategy will not be removed.
+
+
+### Implementation details
+
+* Update APIResource to contain a new member:
```

Comment:
where would this new property be used ?

Reply:
I assume this property to be set to the kubefedcluster or per federated resource. I think I'd go with a simple approach and apply per kubefed cluster so you ensure to delete all the related objects.

Reply:
I agree that it makes sense to have this as a global default also. This could be a property on both the federated cluster and the federated resource, with some strategy like clusters property would be used if available, else look for the one on the resource.